File: Prisma/Code/Src/CSharp/Infrastructure/Python/Wrappers/PrismaOcrWrapper.cs
using CSnakes.Runtime.Python;
using System;
using System.Collections.Generic;

namespace ExxerCube.Prisma.Infrastructure.Python.Wrappers;

/// <summary>
/// CSnakes wrapper for Prisma OCR processing.
/// Provides type-safe Python integration for OCR operations.
/// </summary>
public class PrismaOcrWrapper : IPrismaOcrWrapper, IDisposable
{
    private readonly PyObject _wrapperObject;
    private bool _disposed;

    /// <summary>
    /// Initializes a new instance of the <see cref="PrismaOcrWrapper"/> class.
    /// </summary>
    /// <param name="wrapperObject">The Python wrapper object.</param>
    internal PrismaOcrWrapper(PyObject wrapperObject)
    {
        _wrapperObject = wrapperObject ?? throw new ArgumentNullException(nameof(wrapperObject));
    }

    /// <summary>
    /// Creates a new Prisma OCR wrapper instance.
    /// </summary>
    /// <returns>A new Prisma OCR wrapper.</returns>
    public static PrismaOcrWrapper Create()
    {
        try
        {
            // Get the Python environment and create the wrapper object
            var env = PrismaPythonEnvironment.Env;
            var wrapperObject = env.PrismaOcrWrapper();
            return new PrismaOcrWrapper(wrapperObject);
        }
        catch (Exception ex)
        {
            throw new InvalidOperationException($"Failed to create Prisma OCR wrapper: {ex.Message}", ex);
        }
    }

    /// <summary>
    /// Executes OCR on image data using Python modules.
    /// </summary>
    /// <param name="imageData">The raw image bytes.</param>
    /// <param name="config">The OCR configuration dictionary.</param>
    /// <returns>A dictionary containing OCR results.</returns>
    public Dictionary<string, object> ExecuteOcr(byte[] imageData, Dictionary<string, object> config)
    {
        if (_disposed) throw new ObjectDisposedException(nameof(PrismaOcrWrapper));
        
        try
        {
            // Call the Python execute_ocr function
            var result = _wrapperObject.execute_ocr(imageData, config);
            
            // Convert PyObject result to C# dictionary
            return ConvertPyObjectToDictionary(result);
        }
        catch (Exception ex)
        {
            // Return error result in the expected format
            return new Dictionary<string, object>
            {
                ["error"] = ex.Message,
                ["text"] = "",
                ["confidence_avg"] = 0.0,
                ["confidence_median"] = 0.0,
                ["confidences"] = new List<float>(),
                ["language_used"] = "spa",
                ["processing_errors"] = new List<string> { ex.Message }
            };
        }
    }

    /// <summary>
    /// Extracts structured fields from OCR text.
    /// </summary>
    /// <param name="text">The OCR text content.</param>
    /// <param name="confidence">The OCR confidence score.</param>
    /// <returns>A dictionary containing extracted fields.</returns>
    public Dictionary<string, object> ExtractFieldsFromText(string text, double confidence)
    {
        if (_disposed) throw new ObjectDisposedException(nameof(PrismaOcrWrapper));
        
        try
        {
            // Call the Python extract_fields_from_text function
            var result = _wrapperObject.extract_fields_from_text(text, confidence);
            
            // Convert PyObject result to C# dictionary
            return ConvertPyObjectToDictionary(result);
        }
        catch (Exception ex)
        {
            // Return error result in the expected format
            return new Dictionary<string, object>
            {
                ["expediente"] = null,
                ["causa"] = null,
                ["accion_solicitada"] = null,
                ["fechas"] = new List<string>(),
                ["montos"] = new List<object>(),
                ["confidence"] = confidence,
                ["error"] = ex.Message
            };
        }
    }

    /// <summary>
    /// Extracts expediente (case file number) from text.
    /// </summary>
    /// <param name="text">The text to process.</param>
    /// <returns>The extracted expediente or null.</returns>
    public string? ExtractExpedienteFromText(string text)
    {
        if (_disposed) throw new ObjectDisposedException(nameof(PrismaOcrWrapper));
        
        try
        {
            // Call the Python extract_expediente_from_text function
            var result = _wrapperObject.extract_expediente_from_text(text);
            
            // Convert PyObject result to string or null
            return ConvertPyObjectToString(result);
        }
        catch (Exception)
        {
            return null;
        }
    }

    /// <summary>
    /// Extracts causa (cause) from text.
    /// </summary>
    /// <param name="text">The text to process.</param>
    /// <returns>The extracted causa or null.</returns>
    public string? ExtractCausaFromText(string text)
    {
        if (_disposed) throw new ObjectDisposedException(nameof(PrismaOcrWrapper));
        
        try
        {
            // Call the Python extract_causa_from_text function
            var result = _wrapperObject.extract_causa_from_text(text);
            
            // Convert PyObject result to string or null
            return ConvertPyObjectToString(result);
        }
        catch (Exception)
        {
            return null;
        }
    }

    /// <summary>
    /// Extracts accion solicitada (requested action) from text.
    /// </summary>
    /// <param name="text">The text to process.</param>
    /// <returns>The extracted accion solicitada or null.</returns>
    public string? ExtractAccionSolicitadaFromText(string text)
    {
        if (_disposed) throw new ObjectDisposedException(nameof(PrismaOcrWrapper));
        
        try
        {
            // Call the Python extract_accion_solicitada_from_text function
            var result = _wrapperObject.extract_accion_solicitada_from_text(text);
            
            // Convert PyObject result to string or null
            return ConvertPyObjectToString(result);
        }
        catch (Exception)
        {
            return null;
        }
    }

    /// <summary>
    /// Extracts dates from text.
    /// </summary>
    /// <param name="text">The text to process.</param>
    /// <returns>A list of extracted dates.</returns>
    public List<string> ExtractDatesFromText(string text)
    {
        if (_disposed) throw new ObjectDisposedException(nameof(PrismaOcrWrapper));
        
        try
        {
            // Call the Python extract_dates_from_text function
            var result = _wrapperObject.extract_dates_from_text(text);
            
            // Convert PyObject result to list of strings
            return ConvertPyObjectToListOfStrings(result);
        }
        catch (Exception)
        {
            return new List<string>();
        }
    }

    /// <summary>
    /// Extracts monetary amounts from text.
    /// </summary>
    /// <param name="text">The text to process.</param>
    /// <returns>A list of extracted amounts with currency and value.</returns>
    public List<Dictionary<string, object>> ExtractAmountsFromText(string text)
    {
        if (_disposed) throw new ObjectDisposedException(nameof(PrismaOcrWrapper));
        
        try
        {
            // Call the Python extract_amounts_from_text function
            var result = _wrapperObject.extract_amounts_from_text(text);
            
            // Convert PyObject result to list of dictionaries
            return ConvertPyObjectToListOfDictionaries(result);
        }
        catch (Exception)
        {
            return new List<Dictionary<string, object>>();
        }
    }

    /// <summary>
    /// Converts a PyObject to a C# dictionary.
    /// </summary>
    /// <param name="pyObject">The Python object to convert.</param>
    /// <returns>A C# dictionary representation.</returns>
    private static Dictionary<string, object> ConvertPyObjectToDictionary(PyObject pyObject)
    {
        if (pyObject == null) return new Dictionary<string, object>();
        
        try
        {
            // Use CSnakes built-in conversion
            return pyObject.As<Dictionary<string, object>>();
        }
        catch
        {
            // Fallback: try to convert manually
            var result = new Dictionary<string, object>();
            foreach (var key in pyObject.Keys)
            {
                var value = pyObject[key];
                result[key] = ConvertPyObjectValue(value);
            }
            return result;
        }
    }

    /// <summary>
    /// Converts a PyObject to a string or null.
    /// </summary>
    /// <param name="pyObject">The Python object to convert.</param>
    /// <returns>A string or null.</returns>
    private static string? ConvertPyObjectToString(PyObject pyObject)
    {
        if (pyObject == null) return null;
        
        try
        {
            return pyObject.As<string>();
        }
        catch
        {
            return null;
        }
    }

    /// <summary>
    /// Converts a PyObject to a list of strings.
    /// </summary>
    /// <param name="pyObject">The Python object to convert.</param>
    /// <returns>A list of strings.</returns>
    private static List<string> ConvertPyObjectToListOfStrings(PyObject pyObject)
    {
        if (pyObject == null) return new List<string>();
        
        try
        {
            return pyObject.As<List<string>>();
        }
        catch
        {
            // Fallback: try to convert manually
            var result = new List<string>();
            var length = pyObject.Length;
            for (int i = 0; i < length; i++)
            {
                var item = pyObject[i];
                var str = ConvertPyObjectToString(item);
                if (str != null)
                    result.Add(str);
            }
            return result;
        }
    }

    /// <summary>
    /// Converts a PyObject to a list of dictionaries.
    /// </summary>
    /// <param name="pyObject">The Python object to convert.</param>
    /// <returns>A list of dictionaries.</returns>
    private static List<Dictionary<string, object>> ConvertPyObjectToListOfDictionaries(PyObject pyObject)
    {
        if (pyObject == null) return new List<Dictionary<string, object>>();
        
        try
        {
            return pyObject.As<List<Dictionary<string, object>>>();
        }
        catch
        {
            // Fallback: try to convert manually
            var result = new List<Dictionary<string, object>>();
            var length = pyObject.Length;
            for (int i = 0; i < length; i++)
            {
                var item = pyObject[i];
                var dict = ConvertPyObjectToDictionary(item);
                result.Add(dict);
            }
            return result;
        }
    }

    /// <summary>
    /// Converts a PyObject value to a C# object.
    /// </summary>
    /// <param name="pyObject">The Python object to convert.</param>
    /// <returns>A C# object.</returns>
    private static object ConvertPyObjectValue(PyObject pyObject)
    {
        if (pyObject == null) return null!;
        
        try
        {
            // Try common types
            if (pyObject.IsString) return pyObject.As<string>();
            if (pyObject.IsNumber) return pyObject.As<double>();
            if (pyObject.IsBool) return pyObject.As<bool>();
            if (pyObject.IsList) return ConvertPyObjectToListOfStrings(pyObject);
            if (pyObject.IsDict) return ConvertPyObjectToDictionary(pyObject);
            
            // Default to string representation
            return pyObject.ToString();
        }
        catch
        {
            return pyObject.ToString();
        }
    }

    /// <summary>
    /// Disposes the wrapper and releases Python resources.
    /// </summary>
    public void Dispose()
    {
        if (!_disposed)
        {
            _wrapperObject?.Dispose();
            _disposed = true;
        }
    }


}

File: Prisma/Code/Src/CSharp/Infrastructure/Python/python/prisma_ocr_wrapper.py
#!/usr/bin/env python3
"""
CSnakes-compatible Python wrapper for Prisma OCR processing.
Replaces process-based CLI calls with direct function calls.
"""
import sys
import json
import base64
from pathlib import Path
from typing import Dict, List, Optional, Any, Tuple
import io
from PIL import Image
import numpy as np

# Add the ocr_modules to Python path
sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent.parent.parent.Python))

from ocr_modules import (
    process_path, create_default_config, ProcessingConfig, OCRConfig,
    extract_expediente, extract_causa, extract_accion_solicitada,
    extract_dates, extract_amounts
)

def execute_ocr(image_data: bytes, config: Dict[str, Any]) -> Dict[str, Any]:
    """
    Execute OCR on image data using CSnakes-compatible interface.
    
    Args:
        image_data: Raw image bytes
        config: OCR configuration dictionary
        
    Returns:
        Dictionary containing OCR results
    """
    try:
        # Convert bytes to PIL Image
        image = Image.open(io.BytesIO(image_data))
        
        # Convert config dict to OCRConfig object
        ocr_config = OCRConfig(
            language=config.get('language', 'spa'),
            fallback_language=config.get('fallback_language', 'eng'),
            oem=config.get('oem', 3),
            psm=config.get('psm', 6)
        )
        
        # Create processing config
        processing_config = ProcessingConfig(
            remove_watermark=True,
            deskew=True,
            binarize=True,
            ocr_config=ocr_config,
            extract_sections=True,
            normalize_text=True
        )
        
        # Process the image
        result = process_path(image, processing_config)
        
        # Convert result to dictionary
        return {
            'text': result.ocr_result.text if result.ocr_result else '',
            'confidence_avg': result.ocr_result.confidence_avg if result.ocr_result else 0.0,
            'confidence_median': result.ocr_result.confidence_median if result.ocr_result else 0.0,
            'confidences': result.ocr_result.confidences if result.ocr_result else [],
            'language_used': result.ocr_result.language_used if result.ocr_result else 'spa',
            'processing_errors': result.processing_errors if result.processing_errors else []
        }
        
    except Exception as e:
        return {
            'error': str(e),
            'text': '',
            'confidence_avg': 0.0,
            'confidence_median': 0.0,
            'confidences': [],
            'language_used': 'spa',
            'processing_errors': [str(e)]
        }

def extract_fields_from_text(text: str, confidence: float) -> Dict[str, Any]:
    """
    Extract structured fields from OCR text.
    
    Args:
        text: OCR text content
        confidence: OCR confidence score
        
    Returns:
        Dictionary containing extracted fields
    """
    try:
        # Extract expediente
        expediente = extract_expediente(text)
        
        # Extract causa
        causa = extract_causa(text)
        
        # Extract accion solicitada
        accion_solicitada = extract_accion_solicitada(text)
        
        # Extract dates
        dates = extract_dates(text)
        
        # Extract amounts
        amounts = extract_amounts(text)
        
        return {
            'expediente': expediente,
            'causa': causa,
            'accion_solicitada': accion_solicitada,
            'fechas': dates,
            'montos': amounts,
            'confidence': confidence,
            'error': None
        }
        
    except Exception as e:
        return {
            'expediente': None,
            'causa': None,
            'accion_solicitada': None,
            'fechas': [],
            'montos': [],
            'confidence': confidence,
            'error': str(e)
        }

def extract_expediente_from_text(text: str) -> Optional[str]:
    """
    Extract expediente (case file number) from text.
    
    Args:
        text: The text to process
        
    Returns:
        Extracted expediente or None
    """
    try:
        return extract_expediente(text)
    except Exception:
        return None

def extract_causa_from_text(text: str) -> Optional[str]:
    """
    Extract causa (cause) from text.
    
    Args:
        text: The text to process
        
    Returns:
        Extracted causa or None
    """
    try:
        return extract_causa(text)
    except Exception:
        return None

def extract_accion_solicitada_from_text(text: str) -> Optional[str]:
    """
    Extract accion solicitada (requested action) from text.
    
    Args:
        text: The text to process
        
    Returns:
        Extracted accion solicitada or None
    """
    try:
        return extract_accion_solicitada(text)
    except Exception:
        return None

def extract_dates_from_text(text: str) -> List[str]:
    """
    Extract dates from text.
    
    Args:
        text: The text to process
        
    Returns:
        List of extracted dates
    """
    try:
        return extract_dates(text)
    except Exception:
        return []

def extract_amounts_from_text(text: str) -> List[Dict[str, Any]]:
    """
    Extract monetary amounts from text.
    
    Args:
        text: The text to process
        
    Returns:
        List of extracted amounts with currency and value
    """
    try:
        amounts = extract_amounts(text)
        return [
            {
                'currency': amount.currency,
                'value': float(amount.value),
                'original_text': amount.original_text
            }
            for amount in amounts
        ]
    except Exception:
        return []
